msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java

```diff
@@ -531,57 +525,143 @@ public void skipValue() throws IOException {
      */
     private static MessageTypeException unexpected(String expected, byte b)
             throws MessageTypeException {
-        ValueType type = ValueType.valueOf(b);
-        return new MessageTypeException(String.format("Expected %s, but got %s (%02x)", expected, type.toTypeName(), b));
+        MessageFormat format = MessageFormat.valueOf(b);
+        String typeName;
+        if (format == MessageFormat.NEVER_USED) {
+            typeName = "NeverUsed";
+        } else {
+            String name = format.getValueType().name();
+            typeName = name.substring(0, 1) + name.substring(1).toLowerCase();
+        }
+        return new MessageTypeException(String.format("Expected %s, but got %s (%02x)", expected, typeName, b));
     }
 
-    public MessageFormat unpackValue(ValueHolder holder) throws IOException {
+    public ImmutableValue unpackValue() throws IOException {
         MessageFormat mf = getNextFormat();
         switch(mf.getValueType()) {
             case NIL:
                 unpackNil();
-                holder.setNil();
-                break;
+                return ValueFactory.newNilValue();
             case BOOLEAN:
-                holder.setBoolean(unpackBoolean());
-                break;
-            case INTEGER: {
-                unpackInteger(holder.getIntegerHolder());
-                holder.setToInteger();
-                break;
+                return ValueFactory.newBooleanValue(unpackBoolean());
+            case INTEGER:
+                switch (mf) {
+                    case UINT64:
+                        return ValueFactory.newIntegerValue(unpackBigInteger());
+                    default:
+                        return ValueFactory.newIntegerValue(unpackLong());
+                }
+            case FLOAT:
+                return ValueFactory.newFloatValue(unpackDouble());
```

Comment:
I don't think it's a big deal. I mean we can say that it's a limitation because of internal representation of a Value. Anyways it is impossible to preserve all information of floating point numbers using decimal string format. We should accept that toString of Float values is best-effort.

If we create ImmutableFloatValueImpl,

1) Variable needs another float field which causes overhead and a lot more complexity.
2) Implementation of Value#equals becomes unclear.

I think benefit to have ImmutableFloatValueImpl is smaller than those negative effects.

If you think we should fix it, I suggest following solutions:

c) Always use ImmutableFloatValue always uses Float.toString. Give up double precision when users use json (decimal) string format.
a) Add `private final boolean singlePrecision" field to ImmutableFloatValue (and Variable) class. Only toString method considers it. If it is true, ImmutableFloatValueImpl#toString uses Float.toString.
b) Delete `ValueFactory#newFloatValue(float)` so that users can't create a Value with 0.134f at least in Java.